File: Algorithms/src/3315. Construct the Minimum Bitwise Array II/Code/solution.cpp
class Solution
{
public:
    vector<int> minBitwiseArray(vector<int> &nums)
    {
        for (int &p : nums)
        {
            // Find the bit that can be safely removed
            int removable = ((p + 1) & ~p) >> 1;

            // If removable == 0, number is invalid (even prime like 2)
            // In that case flip all bits to force -1
            p ^= removable | (-(removable == 0) & ~p);
        }
        return nums;
    }
};

File: Algorithms/src/3315. Construct the Minimum Bitwise Array II/Code/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var minBitwiseArray = function (nums) {
  for (let i = 0; i < nums.length; i++) {
    let p = nums[i];

    let removable = ((p + 1) & ~p) >> 1;

    if (removable === 0) {
      nums[i] = -1;
    } else {
      nums[i] = p ^ removable;
    }
  }
  return nums;
};
